crates/turborepo-lib/src/config/override_env.rs

```diff
@@ -0,0 +1,493 @@
+        if input.TURBO_TEAMID.is_some() {
+            output.team_id = input.TURBO_TEAMID;
+            return output;
+        }
+
+        if input.TURBO_TEAM.is_some() {
+            output.team_slug = input.TURBO_TEAM;
+
+            if input.VERCEL_ARTIFACTS_OWNER.is_some() {
+                output.team_id = input.VERCEL_ARTIFACTS_OWNER;
+            }
+
+            return output;
+        }
+
+        if input.VERCEL_ARTIFACTS_OWNER.is_some() {
+            output.team_id = input.VERCEL_ARTIFACTS_OWNER;
+            return output;
+        }
+
+        Output::new()
+    }
+}
+
+pub struct OverrideEnvVars<'a> {
+    environment: &'a HashMap<OsString, OsString>,
+    output: Output,
+}
+
+impl<'a> OverrideEnvVars<'a> {
+    pub fn new(environment: &'a HashMap<OsString, OsString>) -> Result<Self, Error> {
+        let input = Input::from(environment);
+        let output = Output::from(input);
+
+        Ok(Self {
+            environment,
+            output,
+        })
+    }
+
+    fn ui(&self) -> Option<UIMode> {
+        // TODO double check on what's going on here
+        let value = self
+            .environment
+            .get(OsStr::new("ci"))
+            .or_else(|| self.environment.get(OsStr::new("no_color")))?;
+        match truth_env_var(value.to_str()?)? {
+            true => Some(UIMode::Stream),
+            false => None,
+        }
+    }
+}
+
+impl<'a> ResolvedConfigurationOptions for OverrideEnvVars<'a> {
+    fn get_configuration_options(
+        &self,
+        _existing_config: Option<&ConfigurationOptions>,
+    ) -> Result<ConfigurationOptions, Error> {
+        let output = ConfigurationOptions {
+            team_id: self.output.team_id.clone(),
+            token: self.output.token.clone(),
+            team_slug: self.output.team_slug.clone(),
+            ui: self.ui(),
+            api_url: None,
+            ..Default::default()
+        };
+        Ok(output)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use lazy_static::lazy_static;
+
+    use super::*;
+
+    lazy_static! {
+        static ref VERCEL_ARTIFACTS_OWNER: String = String::from("valueof:VERCEL_ARTIFACTS_OWNER");
+        static ref VERCEL_ARTIFACTS_TOKEN: String = String::from("valueof:VERCEL_ARTIFACTS_TOKEN");
+        static ref TURBO_TEAMID: String = String::from("valueof:TURBO_TEAMID");
+        static ref TURBO_TEAM: String = String::from("valueof:TURBO_TEAM");
+        static ref TURBO_TOKEN: String = String::from("valueof:TURBO_TOKEN");
+    }
+
+    struct TestCase {
+        input: Input,
+        output: Output,
+        reason: &'static str,
+    }
+
+    impl TestCase {
+        fn new() -> Self {
+            Self {
+                input: Input::new(),
+                output: Output::new(),
+                reason: "missing",
+            }
+        }
+
+        fn reason(mut self, reason: &'static str) -> Self {
+            self.reason = reason;
+            self
+        }
```

Comment:
Very unimportant since this is test code, but we could make we remember to provide a reason by requiring it in the constructor.
```suggestion
        fn new(reason: &'static str) -> Self {
            Self {
                input: Input::new(),
                output: Output::new(),
                reason,
            }
        }
```

Reply:
so funny you should mention that - that's what I originally had, but `rustfmt` made it really miserable so I switched to this.  if you don't care I'll leave it like this because it means that each case looks the same rather than some of them wrapping and some of them not (including continuations).